FWCore/Framework/src/EventSetupRecordIOVQueue.cc

```diff
@@ -0,0 +1,132 @@
+        if (recordProvider->intervalStatus() == EventSetupRecordProvider::IntervalStatus::NewInterval) {
+          newIntervalForAnySubProcess = true;
+          break;
+        }
+      }
+      for (auto& recordProvider : recordProviders_) {
+        recordProvider->setNewIntervalForAnySubProcess(newIntervalForAnySubProcess);
+      }
+    }
+
+    void EventSetupRecordIOVQueue::checkForNewIOVs(WaitingTaskHolder const& taskToStartAfterIOVInit,
+                                                   WaitingTaskList& endIOVWaitingTasks,
+                                                   bool newEventSetupImpl) {
+      for (auto& recordProvider : recordProviders_) {
+        if (recordProvider->newIntervalForAnySubProcess()) {
+          startNewIOVAsync(taskToStartAfterIOVInit, endIOVWaitingTasks);
+          return;
+        }
+      }
+
+      for (auto& recordProvider : recordProviders_) {
+        if (recordProvider->intervalStatus() != EventSetupRecordProvider::IntervalStatus::Invalid) {
+          endIOVWaitingTasks.add(endIOVWaitingTask_);
+          break;
+        }
+      }
+
+      for (auto& recordProvider : recordProviders_) {
+        recordProvider->continueIOV(newEventSetupImpl);
+      }
+    }
+
+    void EventSetupRecordIOVQueue::startNewIOVAsync(WaitingTaskHolder const& taskToStartAfterIOVInit,
+                                                    WaitingTaskList& endIOVWaitingTasks) {
+      ++cacheIdentifier_;
+      {
+        // Let the old IOV end when all the lumis using it are done.
+        // otherwise we'll deadlock when there is only one thread.
+        WaitingTaskHolder tmp{std::move(endIOVTaskHolder_)};
+      }
+      auto taskHolder = std::make_shared<WaitingTaskHolder>(taskToStartAfterIOVInit);
+      auto startIOVForRecord =
+          [this, taskHolder, &endIOVWaitingTasks](edm::LimitedTaskQueue::Resumer iResumer) mutable {
+            try {
+              unsigned int iovIndex = 0;
+              for (; iovIndex < nConcurrentIOVs_; ++iovIndex) {
+                bool expected = true;
+                if (isAvailable_[iovIndex]->compare_exchange_strong(expected, false)) {
+                  break;
+                }
+              }
+              // Should never fail, just a sanity check
+              if (iovIndex == nConcurrentIOVs_) {
+                throw edm::Exception(edm::errors::LogicError)
+                    << "EventSetupRecordIOVQueue::startNewIOVAsync\n"
+                    << "Couldn't find available IOV slot. This should never happen.\n"
+                    << "Contact a Framework Developer\n";
+              }
+              for (auto recordProvider : recordProviders_) {
+                recordProvider->initializeForNewIOV(iovIndex, cacheIdentifier_);
+              }
+
+              // Needed so the EventSetupRecordIOVQueue destructor knows when
+              // it can run.
+              waitForIOVsInFlight_->increment_ref_count();
```

Comment:
Would it be better to increment this before the lambda is created?

Reply:
I think this increment is in the best place. The lambda created immediately after this line of code is what does the corresponding decrement of this ref count. You only want to increment if you decrement, so it is good to have these as close together as possible. If something happened in between incrementing and decrementing (like an exception being thrown), then that would be a problem because it would hang in an infinite wait in the EventSetupRecordIOVQueue destructor. 

We need this counter for the following reason. We keep the most recent IOV active even when runs and lumis finish, because we do not know when there will be more runs and lumis that want to use the same IOV. We do not know at endLuminosityBlock or endRun whether the IOV will run. The endIOVTaskHolder_ data member of EventSetupRecordIOVQueue is what is preventing the IOV from ending. In the EventSetupRecordIOVQueue destructor that holder is released explicitly. Releasing the holder causes the endIOVTask to be spawned. This task uses EventSetupRecordProvider and if we do not have a wait in the destructor that waits for the endIOV task to complete, then there is a race to see whether the asynchronous task completes first or the EventSetupRecordProvider is destroyed first.

One question we discussed was whether we might want to do this at endJob. We could, but it would be extra work to develop code to do that. My thinking was this, in the existing code we destroy the rest of the EventSetup system in the EventProcessor destructor so it seems logically consistent to destroy EventSetupRecordIOVQueue in the same place.